### ISSUE TITLE
Bug: XAI explanations unreliable for out-of-distribution inputs without validation

### ISSUE THREAD

**Issue by anshul23102** [Saatvik-GT/xaytheon#1041]
## Problem Statement

Xaytheon AI explainability system generates explanations for out-of-distribution inputs without warnings or validation checks. When users input data that looks different from the training data, the model produces confident explanations that are statistically unreliable and potentially dangerous in high-stakes applications (healthcare, finance, lending decisions).

**Business Impact:** Users trust unreliable explanations, make wrong decisions, potential legal liability if decisions cause harm.

## Root Cause Analysis

No distribution shift detection:
1. Model trained on specific data distribution
2. User inputs data outside that distribution
3. Model predicts with confidence on unfamiliar patterns
4. Explanations presented as reliable (they're not)

Example: Model trained on 25-65 age demographic, receives 90-year-old. Explains prediction as if confident, but outside training distribution.

## Solution Overview

**1. Implement Distribution Anomaly Detection**
```python
from sklearn.ensemble import IsolationForest
import numpy as np

# Train on training data distribution
anomaly_detector = IsolationForest(contamination=0.01)
anomaly_detector.fit(training_data)

# Check if new input is in distribution
def is_in_distribution(user_input, threshold=-0.5):
    score = anomaly_detector.predict(user_input)
    return score >= threshold

# Before generating explanation
if not is_in_distribution(user_input):
    return {"warning": "Input outside training distribution. Explanation unreliable."}
```

**2. Compare Distributions**
- Compare input distribution to training distribution
- Flag suspicious inputs
- Calculate distance metrics (Wasserstein, KL divergence)

**3. Confidence Calibration**
- Reduce confidence score for out-of-distribution inputs
- Attach uncertainty estimates
- Only explain when confidence > threshold AND in-distribution

**4. User Notification**
- Warn users: "This explanation may be unreliable (unusual input)"
- Show confidence level
- Suggest collecting more similar data

## Type of Change

- [x] Bug fix (explainability reliability)

## Testing Done

**Test Case 1: Out-of-distribution Detection**
- Train model on ages 25-65
- Input age 90
- Expected: Warning "outside training distribution"
- Actual (Before): Confident explanation given (unreliable)

## Related Issue

XAI reliability, model robustness, user trust in AI.

## Suggested Labels

bug, level:intermediate, gssoc26, xai-safety, distribution-shift, explainability, ai-reliability

**Comment by bashyamharika** [comment]
Hi @anshul23102,

I would like to work on this issue under GSSoC.

This is an important XAI-safety and AI reliability issue, as explanations generated for out-of-distribution (OOD) inputs can be misleading and may reduce user trust in the system, especially in high-impact decision-making scenarios.

I have experience working with machine learning evaluation, data distribution analysis, and model reliability. My approach would be to first analyze the current explanation pipeline and understand how explanations are generated and presented to users.

My plan includes:

* Reviewing the existing prediction and explanation workflow.
* Investigating whether any distribution-shift or OOD detection mechanisms already exist.
* Evaluating suitable approaches for identifying inputs that differ significantly from the training distribution.
* Assessing confidence calibration and uncertainty estimation strategies for explanation generation.
* Designing a validation layer that can flag potentially unreliable explanations before they are shown to users.
* Implementing appropriate user-facing warnings and reliability indicators where necessary.
* Adding automated tests covering in-distribution, out-of-distribution, and edge-case scenarios.
* Documenting the methodology and evaluation metrics used to measure explanation reliability.

I would be happy to perform a comprehensive analysis and contribute a robust solution aligned with the project's XAI-safety and reliability objectives.

Kindly consider assigning this issue to me.

Thank you!

/assign

**Comment by pushtikadia** [comment]
Hi @Saatvik-GT! Processing out-of-distribution inputs without explicit anomaly data-filtering yields highly misleading semantic metrics. I'd love to write a validation layer to check data array bounds before they are fed into the explanation generation pipeline. Please assign this issue to me under GSSoC'26!

**Comment by gayatripadalia** [comment]
Hi @anshul23102,
I worked on this issue in PR #1057.
Implemented:
* Out-of-distribution (OOD) input validation.
* Confidence indicators (`high` / `low`).
* User-facing warnings for unreliable explanations.
* Validation checks before explanation generation.
For testing, the implementation treats values outside the expected range as OOD inputs (e.g., an age-like feature outside `0–100`). Inputs such as `120` trigger a warning and low confidence, while valid inputs such as `50` generate normal explanations with high confidence.
This helps prevent explanations from being presented as fully reliable when inputs fall outside the expected range and improves overall XAI reliability.
Fixes #1041.
